Add Cluster object codegen support

yunhanw-google opened the pull request.

vivien-apple committed:
[YAML] Update chip-tool tests template to generates less redundant code (#9641)

mrjerryjohns commented on zzz_generated/all-clusters-app/zap-generated/cluster-objects.h:
```diff
@@ -0,0 +1,1305 @@
+/*
```
We don't need to generate a unique `cluster-objects.cpp`, and `.h` file for each app. In fact, the whole goal of this cluster object model is to have just 1 version across all apps :) 

That version should be in `zzz_generated/common/zap-generated/`

yunhanw-google replied:
Done, thanks

mrjerryjohns commented on zzz_generated/all-clusters-app/zap-generated/cluster-objects.h:
```diff
@@ -0,0 +1,1305 @@
+namespace NeighborTable {
+enum FieldId
+{
+    kExtAddressFieldId = 0; kAgeFieldId = 1; kRloc16FieldId = 2; kLinkFrameCounterFieldId = 3; kMleFrameCounterFieldId = 4;
```
Formatting issue?

yunhanw-google replied:
Done, thanks

mrjerryjohns commented on zzz_generated/all-clusters-app/zap-generated/cluster-objects.h:
```diff
@@ -0,0 +1,1305 @@
+    uint8_t FabricIndex;
+    chip::ByteSpan RootPublicKey;
+    uint16_t VendorId;
+    FABRIC_ID FabricId;
```
Right type?

yunhanw-google replied:
fixed, thanks

mrjerryjohns commented on zzz_generated/all-clusters-app/zap-generated/cluster-objects.h:
```diff
@@ -0,0 +1,1305 @@
+    chip::ByteSpan RootPublicKey;
+    uint16_t VendorId;
+    FABRIC_ID FabricId;
+    NODE_ID NodeId;
```
Right type?

yunhanw-google replied:
Fixed, thanks

mrjerryjohns commented:
Couple of key items missing:
* Needs to delete the 'fake' clusters-objects I've checked into `src/app/common`
* The BUILD.gn in that folder needs to be updated to pick up the **single** cluster-objects.cpp in zzz-generated/app-common/zap-generated/cluster-objects.cpp` (which is where you'll code-gen to).
* We need to decide if that BUILD.gn should automatically include the `zzz-generated/app-common/zap-generated/` as part of the include path, so that consumers using cluster objects can just do:

```
#include <cluster-objects.h>
```

And not worry about the actual location of the cluster objects.

mrjerryjohns commented:
We also then need to add that BUILD.gn as a dependency for all example apps today to ensure it gets built.

mspang committed:
Fix wrong loop condition in fake MdnsImpl (#9905)

This fixes an out of bounds access in TestPlatform:

==3694438== Conditional jump or move depends on uninitialised value(s)
==3694438==    at 0x4ADA51E: __vfprintf_internal (vfprintf-internal.c:1688)
==3694438==    by 0x114644: chip::Logging::Platform::LogV(char const*, unsigned char, char const*, __va_list_tag*) (Logging.cpp:14)
==3694438==    by 0x11324F: chip::Logging::LogV(unsigned char, unsigned char, char const*, __va_list_tag*) (CHIPLogging.cpp:161)
==3694438==    by 0x1131BB: chip::Logging::Log(unsigned char, unsigned char, char const*, ...) (CHIPLogging.cpp:139)
==3694438==    by 0x113CFF: chip::Mdns::test::CheckExpected(chip::Mdns::test::CallType, chip::Mdns::MdnsService const*) (MdnsImpl.cpp:80)
==3694438==    by 0x113DF7: chip::Mdns::ChipMdnsPublishService(chip::Mdns::MdnsService const*) (MdnsImpl.cpp:105)
==3694438==    by 0x10F647: chip::Mdns::DiscoveryImplPlatform::Advertise(chip::Mdns::OperationalAdvertisingParameters const&) (Discovery_ImplPlatform.cpp:434)
==3694438==    by 0x10B2CA: (anonymous namespace)::TestStub(_nlTestSuite*, void*) (TestPlatform.cpp:161)
==3694438==    by 0x1144AC: nlTestRunner (nlunit-test.c:213)
==3694438==    by 0x10BCFA: TestMdnsPlatform() (TestPlatform.cpp:227)
==3694438==    by 0x10A282: main (TestPlatform.driver.cpp:32)

As this is undefined behavior it may explain the crash others are seeing
with this test.

Damian-Nordic committed:
[controller] Remove legacy command dispatch method (#9842)

* [controller] Remove legacy command dispatch method

Remove "DiscoverAttributes" commands, the last user of
the legacy command sending mechanism, as global commands are
not required by the Matter spec.

* Remove echo client from Android CHIPTool

The echo client used deprecated TempZCL protocol.

yunhanw-google committed:
update xml for struct and enum with cluster id (#9881)

Add cluster id for enum and struct to create relathship between cluster and enum/struct.

Damian-Nordic committed:
[server] Enhance server storage logs (#9886)

Currently, applications print lots of "Retrieved value from
server storage" messages on startup, even if the value was
actually not found in the storage. Make sure the logs are
printed only when a value is really loaded, stored or
deleted.

bzbarsky-apple committed:
Upload diagnostic logs on Darwin if the build+unit-test task fails. (#9891)

This might allow us some slightly better visibility into unit test
crashes on Darwin.

yunhanw-google commented on zzz_generated/app-common/app-common/zap-generated/cluster-objects.h:
```diff
@@ -0,0 +1,1507 @@
+};
+struct DecodableType
+{
+    DataModel::DecodableList<NestedStructList::Type> a;
```
@mrjerryjohns  this line stucks....NestedStructList::Type ==> NestedStructList::DecodableType

mrjerryjohns commented on zzz_generated/app-common/app-common/zap-generated/cluster-objects.cpp:
```diff
@@ -0,0 +1,1027 @@
+    ReturnErrorOnFailure(reader.ExitContainer(outer));
+    return CHIP_NO_ERROR;
+}
+} // namespace ContentLaunchBrandingInformation
```
Add at least 1, preferably 2 newlines between these namespaces.

fessehaeve committed:
Power Source cluster support (#9239)

* Created Power Source cluster XML description

* Added Power Source cluster to zap_cluster_list

* Added Power Source cluster to all-clusters-app

* Added Power Source cluster to controller-clusters

* Added source files for Power Source cluster

* Regenerate all with Power Source cluster changes

* Updated generated files reference path

* Updated generated headers reference in Power Source cluster code

* Restyled Power Source cluster code to make CI happy

* Changed attribute names to CamelCase style

* Regenerate all after attribute names style change to CamelCase

* zap needs poking to regenerate changed attribute names and it's now done.

* Regenerate all after merging with master

* Regenerate all after merging in master.

* Merged master and resolved all conflicts.

* Missed section from merging to master. Fixes issue with generated files.

* Rebase on master and regenerate all.

* Fixed failing Darwin tests and rebase.

Fixed failing Darwin tests as it was caused by none matching all-clusters-app.zap and controller-clusters.zap files.

Rebase on latest ToT

* Another attempt to resolve conflict in endpoint_config.h

sweetymhaiske committed:
Log fix for Mdns (#9918)

fessehaeve committed:
Regenerate after DiscoverAttributes command is removed. (#9925)

turon committed:
[naming] Fix typo in GetExchangeManager method. (#9915)

Commit:
Identify cluster improvments (#9260)

* move identify cluster defintion into seperate xml file

Complies also to current spec.

* add requried identify type to all clusters app

* add identify-cluster.xml to zcl.json

* move clusters/identify to clusters/identify-server

Aligne to other cluster naming

* fix build after move of identify cluster

* rename identify.cpp/h to identify-server.cpp/h

* modernize identify cluster impl.

* manual remove of identify related code

This files are not really updated by the zap tool. So I had to do it manually.

* fix all-clusters-app/esp32

* fix identify in all-cluster-app.zap

* fix ci builds

* fix warnings

* format identify-server.cpp

* add identify to endpoint 1 to make CI happy

* fix compilation of identify-server.cpp

* Restyled by clang-format

* Update src/app/clusters/identify-server/identify-server.cpp

Co-authored-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

* Update src/app/clusters/identify-server/identify-server.cpp

Co-authored-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

* add some documentation for the identify-server impl.

* regen all

* extend identify-server documentation

* Restyled by whitespace

Co-authored-by: Restyled.io <commits@restyled.io>
Co-authored-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

bzbarsky-apple committed:
Rename InteractionModel::ProtocolCode to InteractionModel::Status. (#9907)

* Rename InteractionModel::ProtocolCode to InteractionModel::Status.

In the spec, this is called "status".  Status messages call a protocol-specific
status a "protocol code", but IM is not really using status messages anyway.

* Update generated files

Damian-Nordic committed:
[nrfconnect] Bump nRF Connect SDK version to 1.7.0 (#9919)

* [nrfconnect] Bump nRF Connect SDK version to 1.7.0

New SDK release has been published, so update the supported
version in CHIP. Also, tune the examples' configuration,
for example take advatange of Bluetooth driver optimizations
in the new release.

* [nrfconnect] Add support for DFU over BLE for nRF53 family

Until now only DFU over BLE was only available for
nRF52840 DK.

* Restyled by prettier-markdown

Co-authored-by: Restyled.io <commits@restyled.io>

bzbarsky-apple committed:
Fix intermittent failure in TestReliableMessageProtocol. (#9927)

We are ending up with queued retransmits on both the initiator and
responder.  The responder has a much longer retransmit interval, but
when we sleep for the initiator to retransmit, we can end up sleeping
too long, so the responder retransmits too.  Make the test handle both
cases.

bzbarsky-apple committed:
Fix TLV test to produce more useful failure output. (#9934)

Assertions inside the Get<> function are hard to trace back to the
actual Get<> caller, because they just list the line number of the
Get<> function.  Replace it with a macro, so we get more useful output
on failure.

Co-authored-by: Boris Itkis <boris.itkis@gmail.com>

austinh0 committed:
[Android] Expand Java GN rules to support more complex usecases (#9876)

* Amend Java build rules to support prebuilt JARs and using Java targets in deps.

* Test Java GN build rules

rob-c0der committed:
KVS: Add error check for offset size (#9954)

Check the offset size isn't bigger then the read size and return
error.

mspang committed:
Fix unsolicited announcements with minimal mDNS (#9910)

As of 97d9f7ea3 ("Address 9154 and CM/AC DNS-SD convergence (#9348)")
operational discovery during chip-tool's commissioning flow typically
fails with minimal mDNS. This occurs because this change unintentionally
broke the logic to send unsolicited mDNS announcements as required by
RFC 6762.

During commissioning, the initial mDNS query to resolve the operational
service on the network is generally sent before the device is able to
join the network. The is perfectly fine because RFC 6762 accounts for
this case and requires devices joining the network to advertise all of
their records. These unsolicited responses provide the IP address and
allow the final commissioning steps to execute.

The minimal mDNS implementation sends its unsolicited announcements in
Start() and not at any other time. In the above change, we started
calling StopPublishDevice() immediately prior to (re-)starting the
server, leaving no records to announce. The records are added
immediately afterwards, but currently adding records does not trigger
additional announcements. Thus, a response never arrives and
commissioning times out.

Fix this by adding an announcement after the records are added.
Unfortunately this will quite spammy since we announce everything and
advertise is called several times in a row. What should happen is
announcements only happen with new or updated records, or when new
interfaces are enabled.

(Note that this problem is greatly exacerbated by the lack of
retransmissions of queries, which is being addressed in
https://github.com/project-chip/connectedhomeip/pull/9900).

pan-apple committed:
Do not disable BLE if a fabric has not been provisioned on the device (#9933)

* Check if a fabric has been provisioned before disabling BLE

* Update src/app/server/Server.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

* check for network configuration as well

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

cecille committed:
Remove ip cluster flags (#9892)

* Remove IP commissioning flags.

We've been running with these default on for a while, we don't
need to be gating on this anymore and it looks to be messing up
folks using chip-tool.

* Don't early return on IP commissioning.

When we're doing full commissioning, we don't want to call the
completion callback until we're actually done. CASE causes the
address to be resolved multiple time, including after we've done
our resolve.

* Restyled by shfmt

Co-authored-by: Restyled.io <commits@restyled.io>

andy31415 committed:
Add support for arm64 RPI cross compile in build_examples.py (#9866)

* Add arm64 cross compile support for linux apps.

Tested with:

```
./scripts/build/build_examples.py --platform host --board arm64 build
```

which generated cross compiled chip tool, thermostat app and
all-clusters app.

* update test runner a bit

* Fix script to run for native, fix unit tests

* Restyle fixes

* Code review updates

* Typo fix

* Restyle fixes

* Update unit test

* Fix up host/board naming for native. Makes away with all -native bits

* Fix unit tests after native board fixup

* Make unit tests for darwin pass. Cross compile for darwin still feels off and unit tests may not pass on RPI because not all targets are supported

yunhanw-google committed:
Update xml (#9912)

* update xml

--update xml with array for struct item
--update InterfaceType for NetworkInterfaceType struct

* run codegen

yunhanw-google committed:
add initial codegen support for cluster object

yunhanw-google committed:
run codegen
